pkg/hubble/filters/protocol.go

```diff
@@ -27,6 +27,10 @@ func filterByProtocol(protocols []string) (FilterFunc, error) {
 	var l4Protocols, l7Protocols []string
 	for _, p := range protocols {
 		proto := strings.ToLower(p)
+		if strings.HasPrefix(proto, "tcp:") {
```

Comment:
I think we should add TCP filtering as a new field to the `FlowFilter` message. I'm not a huge fan of the string-based API here (again, how it will look in the CLI is a separate discussion), as it is prone to errors. For example a typo in a filter `tcp:SYN,AKC` would silently match every flow.

In general, I imagine something dedicated for TCP flags:

### API Proposal

I propose we add a new field to the `FlowFilter` message: https://github.com/cilium/cilium/blob/c142e5e2d0b92686950e553dca5805056fa2007f/api/v1/flow/flow.proto#L249-L251

For example:

```protobuf
message FlowFilter {
   // ...
   repeated TCPFlags tcp_flags = 21;
}

// TCPFlags is already defined as follows:
```protobuf
message TCPFlags {
    bool FIN = 1;
    bool SYN = 2;
    bool RST = 3;
    bool PSH = 4;
    bool ACK = 5;
    bool URG = 6;
    bool ECE = 7;
    bool CWR = 8;
    bool NS = 9;
}
```

Instead of parsing a string in the filter implementation, you would receive one (or more) sets of TCP flags which have to be set for a packet to match the filter. The filter semantics would be the same as you have already implemented it:

For a given `TCPFlags` set, all flags set in the filter must also be set in the flow. Example:
```js
tcp_flags = [{ACK:true, SYN:true}] // matches any flow which have at least both the SYN and ACK flag set
tcp_flags = [{ACK:true},{SYN:true}] // matches any flow which has either the SYN and ACK flag set
```

Using the inclusion and exclusion filter (the `whitelist/blacklist` fields in `GetFlowsRequest`)) we have, one will also be able to exclude flows based on flags:
```js
include:  [{ACK:true, SYN:true}], exclude: [{RST:true},{FIN:true},{URG:true},{CWR: true},{PSH: true},{NS: true},{ECE:true}] // matches any flows which ONLY have the SYN and ACK flag set 
```

### Implementation Step Proposal

1. Add a new `repeated TCPFlags tcp_flags = 21;` to `message FilterFlow`, regenerate the gRPC files (`make -C api/v1`). Commit the generated files.
2. Add a new `pkg/hubble/filters/tcp.go` file, with roughly the following functions (feel free to name things differently of course):

```go
func filterByTCPFlags(flags []*flowpb.TCPFlags) FilterFunc {
   return func(ev *v1.Event) bool {
      // similar implementation as in this PR,
      // but using the array of `TCPFlags` instead of splitting strings
   }
}

type TCPFilter struct{}

// OnBuildFilter builds an TCP filter
func (f *TCPFilter) OnBuildFilter(ctx context.Context, ff *flowpb.FlowFilter) ([]FilterFunc, error) {
	// this is generic in case we will add more TCP related filters in the future
	var fs []FilterFunc 
        // this is the new field added in the proto
	if ff.GetTCPFlags() != nil {
		fs = append(fs, filterByTCPFlags(ff.GetTCPFlags()))
	}
	return fs, nil
}
```

3. Add the `TCPFilter` to this list here:
https://github.com/cilium/cilium/blob/c142e5e2d0b92686950e553dca5805056fa2007f/pkg/hubble/filters/filters.go#L137-L150

Reply:
Hey @gandro, thanks for laying out the detailed picture of the implementation.

One major stumbling block I faced was that I was using an old version of `protoc-gen-go` and it was auto-generating the files *.pb.go in a very different way. The makefile didn't report any problem (makefile currently only verifies `protoc` version and ignores `protoc-gen-go`). I spent a lot of time on this problem and eventually saw the header of existing auto-generated flow.pb.go and realized that specific `protoc-gen-go v1.25.0` needs to be used. Fixing that solved the problem. Note that I am using vagrant based dev setup. Shouldn't the verification of `protoc-gen-go` also be added to api/v1/Makefile?

Anyways, there is one more point I wish to solicit your advice. After adding `tcp.go`, currently, I am testing by adding test code `tcp_test.go`. But shouldn't this TCP filter also show up as a new filter in `hubble observe --help` list of filters? What is the way to make the tcp filter appear in that command line option? Somehow, I couldn't figure this out!

Reply:
@nyrahul  Sorry you spent time on this issue. We actually know that `protoc-gen-go` is missing but this problem is going to be addressed by #12931 so that we have a very repeatable way of generating files from proto definition. Feel free to reach out on [#development](https://cilium.slack.com/archives/C2B917YHE), if you get stuck or have any question.

> Anyways, there is one more point I wish to solicit your advice. After adding tcp.go, currently, I am testing by adding test code tcp_test.go. But shouldn't this TCP filter also show up as a new filter in hubble observe --help list of filters? What is the way to make the tcp filter appear in that command line option? Somehow, I couldn't figure this out!

The CLI changes need to be done in the [hubble CLI repository](https://github.com/cilium/hubble). You can have a look at [`cmd/observe/observe_filter.go`](https://github.com/cilium/hubble/blob/master/cmd/observe/observe_filter.go) to see how the filters are built. And look at [`cmd/observe/observe.go`](https://github.com/cilium/hubble/blob/master/cmd/observe/observe.go) towards the bottom to understand how they're applied to the `GetFlows` request. I wouldn't worry about this at the moment as this can be easily addressed once the server-side changes are in.